sstables/storage.cc

```diff
@@ -41,6 +41,7 @@ namespace sstables {
 class filesystem_storage final : public sstables::storage {
     std::filesystem::path _dir;
     std::optional<std::filesystem::path> _temp_dir; // Valid while the sstable is being created, until sealed
+    mutable std::unordered_map<std::string, file> _cached_dirs;
```

Comment:
Why do we need a cache? Can't an individual owner of an opened directory keep it in a variable?

Reply:
So yes, this is the foundational question. I asked a similar question [here](https://github.com/scylladb/scylladb/issues/2399#issuecomment-2192237504). There are multiple possible levels for directory reuse (for flushing purposes):

- open, flush, and close every single time (what most of the `filesystem_storage` methods do now)
- open, flush several times, then close, in a single method (what `filesystem_storage::seal` does, as of current master)
- try to cache directories by _purpose_ (for example, for all those locations where `_dir.native()` is opened just for the sake of flushing, introduce a new, named data member of type `file`, and flush through that). AIUI, this is what you are referring to, by _individual owner of an opened directory keep it in a variable_. This approach sounds very challenging to me, because I can't readily determine the very first use of `_dir.native()` for this particular purpose (i.e., I can't easily tell who the _individual owner **is**_, for each potentially cached directory), so I'm unsure where to actually initialize each such new data member. Benny explained a number of individual cases [here](https://github.com/scylladb/scylladb/issues/2399#issuecomment-2193823969), but still...
- and finally, a cache that keeps open directories by name (for the sake of flushing), regardless of call site. Very convenient, because I need not collect where each directory is opened first. This was recommended by Benny at the end of his comment [here](https://github.com/scylladb/scylladb/issues/2399#issuecomment-2193823969).

Reply:
The fourth option is just making the code more complex. Yes, it requires less diligence, but it also leaks descriptors.

I don't see why there's be a question about the lifetime of the file member. When you create the `table` object, create that member. When you close the table object, close the member. Pass it around to anyone who needs it.

There is some complexity because of the indirection introduced by sstables::storage. So we can't store a plain file. We could add `future<> storage::synchronize() = 0`. So whereever you have access to storage (should be everywhere?) you can synchronize the directory. @xemul please check the above.

Reply:
I don't think this is a problem at all. Whenever an sstable is opened, its kernel file object holds dentry (and inode) and the dentry itself holds parent dentry in cache. So opening a directory for sync always happens from dcache and doesn't do any IO (we're only interested in XFS case). Am I missing something?

Yes, keeping a file descriptor speeds things up a little bit, but does the saving worth it?

Reply:
> I don't see why there's be a question about the lifetime of the file member. When you create the `table` object, create that member. When you close the table object, close the member. Pass it around to anyone who needs it.

`sync_directory` seems to be called on several more directories than just the singular `_dir` object that seems to underlie (?) the associated (?) `sstable` object.

> There is some complexity because of the indirection introduced by sstables::storage. So we can't store a plain file. We could add `future<> storage::synchronize() = 0`. So whereever you have access to storage (should be everywhere?) you can synchronize the directory.

Apologies, I don't understand this. I assume the abstract / virtual `storage::synchronize()` method would be overridden in both `s3_storage` (to do nothing) and in `filesystem_storage` (to wrap `sync_directory`). However, this would require _lifting_ the `sync_directory` calls from `filesystem_storage` internals, to `sstable` level as `_storage->synchronize()` calls.

I don't see how that could work. First, it looks like a layering violation.

Second, more practically, the current `sync_directory` calls in `filesystem_storage` are intermixed with a bunch of `filesystem_storage`-specific logic (at strategic points). `sstable` code could only call `_storage->synchronize()` before and after other `storage` method calls, not _inside_ them.

Sorry about my confusion.

Reply:
> I don't think this is a problem at all. Whenever an sstable is opened, its kernel file object holds dentry (and inode) and the dentry itself holds parent dentry in cache. So opening a directory for sync always happens from dcache and doesn't do any IO (we're only interested in XFS case). Am I missing something?

This is more of a good-first-issue to get started than solving a pain. Which is why I don't like adding a cache for it, the cure is worse than the disease.

The cost of the open() is a half-millisecond of latency to get to the syscall thread (or io_uring when we finally enable it). It's not a big issue by any means.


> 
> Yes, keeping a file descriptor speeds things up a little bit, but does the saving worth it?

If it's keeping a file object near the directory name, yes, if it's adding a cache + cache management + solving the bugs it raises, no.

Reply:
> > I don't see why there's be a question about the lifetime of the file member. When you create the `table` object, create that member. When you close the table object, close the member. Pass it around to anyone who needs it.
> 
> `sync_directory` seems to be called on several more directories than just the singular `_dir` object that seems to underlie (?) the associated (?) `sstable` object.
> 
> > There is some complexity because of the indirection introduced by sstables::storage. So we can't store a plain file. We could add `future<> storage::synchronize() = 0`. So whereever you have access to storage (should be everywhere?) you can synchronize the directory.
> 
> Apologies, I don't understand this. I assume the abstract / virtual `storage::synchronize()` method would be overridden in both `s3_storage` (to do nothing) and in `filesystem_storage` (to wrap `sync_directory`). However, this would require _lifting_ the `sync_directory` calls from `filesystem_storage` internals, to `sstable` level as `_storage->synchronize()` calls.
> 
> I don't see how that could work. First, it looks like a layering violation.
> 
> Second, more practically, the current `sync_directory` calls in `filesystem_storage` are intermixed with a bunch of `filesystem_storage`-specific logic (at strategic points). `sstable` code could only call `_storage->synchronize()` before and after other `storage` method calls, not _inside_ them.
> 
> Sorry about my confusion.

My memory predates sstables::storage so what I wrote may have been incoherent.

filesystem_storage represents a directory (or perhaps a few). Let's look at filesystem_storage::open():

```c++
void filesystem_storage::open(sstable& sst) {
    touch_temp_dir(sst).get();
...
    sst.sstable_write_io_check(sync_directory, _dir.native()).get();
}
```

If we add alongside _dir a member _dir_file then we can avoid the open(). Similar analysis should follow for the other callers. If one or two turns out to be insurmountable (don't see why), we can keep it using the name.

Reply:
> If we add alongside _dir a member _dir_file then we can avoid the open(). Similar analysis should follow for the other callers.

I've tried to address this in v4 the best I could.